Initial implementation of Map Zones

aleos89 opened the pull request:
* **Addressed Issue(s)**: N/A

* **Server Mode**: Pre-renewal and Renewal

* **Description of Pull Request**: 
  * Implements Map Zones which allows easier configuration of map settings.
  * The restricted mapflag has been removed as they will no longer be needed.
  * General cleanups and optimizations.

aleos89 committed:
Initial implementation of Map Zones
* Implements Map Zones which allows easier configuration of map settings.
* The restricted mapflag has been removed as they will no longer be needed.
* General cleanups and optimizations.

aleos89 committed:
Fixes CI errors

aleos89 committed:
More battle_config cleanups

aleos89 committed:
Fixes PK damage calculation
* Convert a few more PK damage checks that got missed from a merge conflict.

aleos89 committed:
Update messages for other languages

aleos89 committed:
Fix Russian specifier to integer

aleos89 committed:
Disable mapflags on custom scripts

Lemongrass3110 commented on src/map/mob.cpp:
```diff
@@ -2629,10 +2629,12 @@ int mob_dead(struct mob_data *md, struct block_list *src, int type)
 		}
 	}
 
+	map_data *mapdata = map_getmapdata(m);
```
Not sure, but all the changes in this file look like unrelated misc fixes/cleanups?

Lemongrass3110 commented on src/map/pc.cpp:
```diff
@@ -773,6 +776,9 @@ static TIMER_FUNC(pc_invincible_timer){
 void pc_setinvincibletimer(map_session_data* sd, int val) {
 	nullpo_retv(sd);
 
+	if (val <= 0)
```
Maybe still check if the timer needs to be deleted?

Lemongrass3110 commented on src/map/status.cpp:
```diff
@@ -1945,6 +1940,15 @@ bool status_check_skilluse(struct block_list *src, struct block_list *target, ui
+		map_data *mapdata = map_getmapdata(src->m);
+		map_session_data *sd = (TBL_PC *)src;
+
+		if (mapdata != nullptr && mapdata->zone.isSkillDisabled(skill_id, src->type, (sd != nullptr) ? pc_get_group_level(sd) : 0 )) {
```
Below you use 101?

Lemongrass3110 commented on src/map/map.hpp:
```diff
@@ -794,6 +806,90 @@ struct iwall_data {
+		if (this->disabled_commands.empty())
+			return false;
+
+		auto cmd_lv = util::umap_find(this->disabled_commands, name);
```
Based on the calls to this and the other functions below:
My suggestion would be to hand "sd" over and check if it is nullptr and do the according checks in here.

Lemongrass3110 reviewed (commented):
Some initial comments

aleos89 committed:
Translate new messages to Spanish
Thanks to @SapitoSucio!

aleos89 committed:
Implements mapflag MF_FLEE_PENALTY
* Deprecate the bg_flee_penalty and gvg_flee_penalty battle configs.
* Removes hardcoded restrictions limiting to only BG and GvG maps.

aleos89 committed:
Cleanup atcommand mapinfo
* Adjust checks for faster lookup.

aleos89 committed:
Fixes incorrect message display

aleos89 committed:
Remove deprecated databases from file move

aleos89 committed:
Remove plurality

aleos89 committed:
Adds a function to assist with PK damage
* Re-implement battle_calc_pk_damage to help determine when to adjust damage when PK mode is enabled.
* Implement INF2_IGNOREPKREDUCTION skill flag to make skills bypass the damage adjustments.

aleos89 committed:
Fixes invincible timer check

aleos89 committed:
Map Zone parsing optimizations

aleos89 committed:
Invincible timer clean
* Remove a possible timer even if passing an invalid time value.

aleos89 committed:
Adds object type check to isStatusDisabled
* Check by object type rather than passing arbitrary integers to compare against.

Atemo commented on src/map/map.cpp:
```diff
@@ -194,6 +193,312 @@ struct s_generator_options {
+					ShowError("MapZoneDatabase::loadingFinished: Invalid mapflag %s from zone %s.\n", flag_name, script_get_constant_str("MAPTYPE_", zone.first));
+			}
+		}
+	}
```
Maybe we could check all zones (of all maps defined in `map_index.txt`) and add a warning when no zones are defined. This could help with map management.

`ShowWarning("MapZoneDatabase::loadingFinished: Map %s has no zone. Defaults to DUMMY_ZONE.\n", ...);`
(`DUMMY_ZONE` would have no restrictions)

laziem commented:
Can something like these be added?
- OnlyEnableItems
- OnlyEnableSkill
- OnlyEnableStatues

It will be much easier to enable set of items only (ex: Booster equipment)
Will be useful for event or other scenario, if we want player to be able to use certain items/skill on selected map.

Pokye commented on conf/msg_conf/map_msg_por.conf:
```diff
@@ -1052,9 +1052,9 @@
 1038: Digite pelo menos um n�mero v�lido da lista (uso: @mapinfo <0-3> <mapa>).
 1039: ------ Informa��o do Mapa ------
-1040: Mapa: %s | Jogadores: %d | NPCs: %d | Chats: %d | Vendas: %d
+1040: Mapa: %s (Zone: %s) | Jogadores: %d | NPCs: %d | Chats: %d | Vendas: %d
```
```suggestion
1040: Mapa: %s (Zona: %s) | Jogadores: %d | NPCs: %d | Chats: %d | Vendas: %d
```

aleos89 committed:
Adds ability to clear mapflags
* Added the Clear label to allow mapflags to be removed for importing use.
* Converts mapflag storage into multimap.

aleos89 committed:
Convert s_map_zone_data storage to shared_ptr

aleos89 committed:
Fixes mapflag parsing
* Adds back an accidental removal for mapflag name lookup.

aleos89 committed:
Fixes MF_FLEE_PENALTY parsing

aleos89 committed:
Add empty zone notice
* Display a notice when a map doesn't have a zone with a suggestion to assign one.
* Avoiding a default zone as many maps are added from updates and may not get their proper zones until the requisite data is found.

aleos89 committed:
Adds translation for Portuguese
Thanks to @Pokye!

aleos89 committed:
Remove unused battle configs

aleos89 committed:
Fixes atcommands for PvP and GvG toggle
* Fixes atcommands pvpon, pvpoff, gvgon, and gvgoff to properly toggle the current map with the appropriate zone.
* Atcommand pvpon will assign the MAPTYPE_NOPENALTY_FREEPKZONE zone.
* Atcommand gvgon will assign the MAPTYPE_EVENT_GUILDWAR zone.
* Adds getMapZone() to return a map's zone.
* Adds setZone() to set a map's zone.
Thanks to @Atemo!

aleos89 committed:
Revert PvP/GvG atcommands
* Revert atcommand pvpon/pvpoff.
* Revert atcommand gvgon/gvgoff.

aleos89 commented:
> Can something like these be added?
> 
> - OnlyEnableItems
> 
> - OnlyEnableSkill
> 
> - OnlyEnableStatues
> 
> 
> 
> It will be much easier to enable set of items only (ex: Booster equipment)
> 
> Will be useful for event or other scenario, if we want player to be able to use certain items/skill on selected map.

I'll gladly accept this as a feature request after the initial implementation gets put in just to keep "custom" parts of this separate.

Atemo committed:
Maps in order

Atemo committed:
Corrected the zone of few maps

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Resolve compile errors and warnings

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Adjust structs to classes
* Adjust to class to avoid duplication of definitions.
* Rename prefix to match "class".

aleos89 committed:
Refactor a few functions to use sd lookup

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+			}
+
+			if (group_lv == 0)
+				zone->disabled_commands.erase(command_name);
```
Report if trying to erase something not existant?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+
+		for (const auto &it : commandNode) {
+			std::string command_name;
+			c4::from_chars(it.key(), &command_name);
```
should we uppercase/lowercase here?
dont know if umap is case sensitive

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+				std::string bl_name;
+				c4::from_chars(subBl.key(), &bl_name);
+
+				if (bl_name.compare("Skill") == 0)
```
Just because i had to think a little bit to understand "why" this check is here:
Would it be better to remove this string check (that may be case sensitive) and add a BL_SKILL check below the BL_NUL check instead?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+					group_lv = 100;
+				}
+
+				if (util::umap_exists(zone->disabled_skills, skill_id)) {
```
Shouldnt the outer "if" be the group_lv check?
To know if you want to enable or disable?

Again: Report if trying to disable something that is not enabled?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+			}
+
+			if (group_lv == 0)
+				zone->disabled_items.erase(item->nameid);
```
Report if trying to erase something that does not exist?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+			}
+
+			if (group_lv == 0)
+				zone->disabled_statuses.erase(static_cast<sc_type>(status));
```
Report if trying to erase something that does not exist?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+			}
+
+			if (group_lv == 0)
+				zone->restricted_jobs.erase(static_cast<uint32>(job_id));
```
Report if trying to erase something that does not exist?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+				auto mapflagit = zone->mapflags.equal_range(static_cast<int16>(flag));
+
+				// Iterate over the mapflags to get their values
+				for (std::multimap<int16, std::string>::iterator it = mapflagit.first; it != mapflagit.second; ++it) {
```
```suggestion
				for (auto it = mapflagit.first; it != mapflagit.second; ++it) {
```
But I think "it" shadows the loop variable of the outer loop here.

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+	// Check for maps with no zones
+	for (int i = 0; i < map_num; i++) {
+		if (map[i].zone == nullptr)
+			ShowNotice("MapZoneDatabase::loadingFinished: Map %s has no zone. Assigning a zone is highly encouraged.\n", map[i].name);
```
```suggestion
			// TODO: Change to ShowWarning as soon as possible.
			ShowNotice("MapZoneDatabase::loadingFinished: Map %s has no zone. Assigning a zone is highly encouraged.\n", map[i].name);
```

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+		return false;
+	}
+
+	auto zonedata = map_zone_db.find(zone);
```
```suggestion
	auto zonedata = this->find(zone);
```

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -194,6 +194,507 @@ struct s_generator_options {
+
+	// Apply mapflags from Map Zone DB
+	for (const auto &flag : zonedata->mapflags) {
+		char flag_name[50] = {}, empty[1] = {};
```
```suggestion
		char flag_name[100], *empty = "";
```

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -3621,8 +4121,8 @@ void map_flags_init(void){
 		// adjustments
-		if( battle_config.pk_mode && !mapdata_flag_vs2(mapdata) )
-			mapdata->setMapFlag(MF_PVP, true); // make all maps pvp for pk_mode [Valaris]
+		//if( battle_config.pk_mode && !mapdata_flag_vs2(mapdata) ) // !TODO: Is this needed now that the PK zone exists?
```
Open TODO?

Lemongrass3110 commented on src/map/map.cpp:
```diff
@@ -3606,7 +4107,6 @@ void map_flags_init(void){
 		args.flag_val = 100;
 
 		// additional mapflag data
-		mapdata->zone = 0; // restricted mapflag zone
```
mapdata->zone = nullptr;

Lemongrass3110 commented on src/map/battle.cpp:
```diff
@@ -9490,8 +9478,8 @@ int64 battle_calc_return_damage(struct block_list* tbl, struct block_list *src,
-	else if (mapdata->getMapFlag(MF_PVP))
-		rdamage = battle_calc_pk_damage(*src, *tbl, rdamage, skill_id, flag);
+	else if (battle_config.pk_mode > 0 && mapdata->getMapFlag(MF_PVP))
+			damage = battle_calc_pk_damage(*src, *tbl, damage, skill_id, flag);
```
```suggestion
		damage = battle_calc_pk_damage(*src, *tbl, damage, skill_id, flag);
```
Why the change from rdamage to damage?

Lemongrass3110 commented on src/map/atcommand.cpp:
```diff
@@ -4418,23 +4423,17 @@ ACMD_FUNC(mapinfo) {
 
 	struct map_data *mapdata = map_getmapdata(m_id);
```
map_data should be checked for null once here

aleos89 committed:
Clean ups from review
Thanks to @Lemongrass3110!

aleos89 committed:
Fixes a compile error

aleos89 committed:
Merge branch 'master' into feature/map_zones

# Conflicts:
#	conf/msg_conf/map_msg.conf

aleos89 committed:
Cleanups and optimizations

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Resolve map-server-generator post build tasks

aleos89 committed:
Adjust getMapZone
* Change the lookup method for getMapZone to utilize the value stored in the map_data rather than looking through map_zone memory bringing faster results.

Daegaladh commented on conf/msg_conf/map_msg.conf:
```diff
@@ -936,7 +936,11 @@
+833: This command is disabled on this map.
+834: Map Zone database has been reloaded.
+
+//834-899 free
```
```suggestion
//835-899 free
```

Daegaladh commented on conf/msg_conf/map_msg_chn.conf:
```diff
@@ -871,9 +871,9 @@
+1040: 地圖名: %s (Zone: %s) | 地圖玩家數量: %d | 地圖NPC數量: %d | 地圖聊天室數量: %d | Vendings: %d
 1041: ------ 地圖旗標 ------
-1042: 城市地圖
+1042: Battle Rate Adjustments:
```
Why is battle instead of damage?

aleos89 replied:
I guess I was thinking overall "battle" adjustments but can adjust to damage to be a bit more clear!

Daegaladh replied:
I mean because you used 'damage' on the english file

Daegaladh commented on conf/msg_conf/map_msg_frn.conf:
```diff
@@ -883,9 +883,9 @@
+1040: Nom de la Map: %s (Zone: %s) | Joueurs sur la Map: %d | NPCs sur la Map: %d | Chats sur la Map: %d | Vendings: %d
 1041: ------ Map Flags ------
-1042: Town Map
+1042: Battle Rate Adjustments:
```
Again

Daegaladh commented on conf/msg_conf/map_msg_idn.conf:
```diff
@@ -970,9 +970,9 @@
+1040: Map: %s (Zone: %s) | Pemain: %d | NPC: %d | Ruang Chat: %d | Toko: %d
 1041: ------ Map Flags ------
-1042: Kota
+1042: Battle Rate Adjustments:
```
Again

Daegaladh commented on conf/msg_conf/map_msg_rus.conf:
```diff
@@ -884,9 +884,9 @@
+1040: �������� �������: %s (Zone: %s) | ������: %d | ���: %d | ����: %d | Vendings: %d
 1041: ------ �������� ------
-1042: �������
+1042: Battle Rate Adjustments:
```
Again

Daegaladh commented on conf/msg_conf/map_msg_tha.conf:
```diff
@@ -877,9 +877,9 @@
 1041: ------ Map Flags ------
-1042: Town Map
+1042: Battle Rate Adjustments:
 1043: Autotrade Enabled
```
Again

Daegaladh commented on conf/msg_conf/map_msg_spn.conf:
```diff
@@ -1124,9 +1124,9 @@
+1040: Mapa: %s (Zona: %s) | Jugadores: %d | NPCs: %d | Chats: %d | Tiendas: %d
 1041: ------ Opciones del mapa ------
-1042: Mapa de ciudad
+1042: Ajuste Tasa de Batalla:
```
```suggestion
1042: Ajuste de ratios de daño:
```

Daegaladh reviewed (commented):
Just reviewed a few files for now XD

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      prt_castle: true
+      prt_church: true
+      prt_in: true
+      prt_lib: true
+      prt_lib_q: true
+      prt_pri00: true
+      que_house_s: true
```
```suggestion
```
Not pre-re

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      ra_in01: true
+      ra_temin: true
+      ra_temsky: true
+      rebel_in: true
```
```suggestion
```
Not pre-re

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      ra_temsky: true
+      rebel_in: true
+      s_atelier: true
+      sch_lab: true
```
This map isn't on any database, I think is custom from jRO or something

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      ra_temin: true
+      ra_temsky: true
+      rebel_in: true
+      s_atelier: true
```
This should go after spl_in02

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      sec_in01: true
+      spl_in01: true
+      spl_in02: true
+      star_in: true
```
```suggestion
```
Not pre-re

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      star_in: true
+      turbo_room: true
+      um_in: true
+      un_bk_q: true
```
```suggestion
```
Not pre-re

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      un_bk_q: true
+      ve_in: true
+      ve_in02: true
+      verus04: true
```
```suggestion
```
Not pre-re

Daegaladh commented on db/pre-re/map_zones.yml:
```diff
@@ -0,0 +1,1773 @@
+      ve_in02: true
+      verus04: true
+      xmas_in: true
+      y_airport: true
+      yuno_in01: true
+      yuno_in02: true
+      yuno_in03: true
+      yuno_in04: true
+      yuno_in05: true
+      yuno_pre: true
```
```suggestion
      yuno_in01: true
      yuno_in02: true
      yuno_in03: true
      yuno_in04: true
      yuno_in05: true
      yuno_pre: true
      y_airport: true
```
Alphabetical

Daegaladh reviewed (commented):
Just reviewed Village and Village_In for now

Daegaladh committed:
Fixed alphabetical order and remove some renewal maps from pre-e

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Resolve a merge error

aleos89 committed:
Adjust pre-re map zones
Thanks to @Daegaladh!

aleos89 committed:
Minor message cleanups
Thanks to @Daegaladh!

Daegaladh committed:
Fixed pre-re maps
Removed renewal maps from pre-re
Improved map descriptions
Fixed pre-re bossnia maptype (It's actually 24)

Haydrich commented:
I've test it, i just got some issues, Some Ids is missing the Invincible_Time, for example Id: Dungeon
another thing it forget the command @reloadzonedb

```
		ACMD_DEF2("reloadzonedb",reload),
```

aleos89 committed:
Add missing atcommand reference

aleos89 committed:
Add back player_invincible_time battle config
* Resolves an issue where the invincible time as not getting applied to maps without the MF_INVINCIBLE_TIME flag.
* Move the logic checks for MF_INVINCIBLE_TIME, MF_FLEE_PENALTY, MF_WEAPON_DAMAGE_RATE, MF_MAGIC_DAMAGE_RATE, MF_MISC_DAMAGE_RATE, MF_LONG_DAMAGE_RATE, and MF_SHORT_DAMAGE_RATE to map_setmapflag_sub() from npc_parse_mapflag().
* Update the mapflag documentation.

aleos89 committed:
Merge branch 'master' into feature/map_zones

aleos89 committed:
Adjust map ID lookup to be primary
* Resolves a few cases of where a map may be defined in the index/cache but not enabled from the config.
* Comment out a few pre-renewal maps that are not enabled.

aleos89 committed:
Add some renewal maps to zones

aleos89 committed:
Add inheritance support
* Disabled commands, disabled skills, disabled items, disabled statuses, restricted jobs, and mapflags can be inherited from other zones.

aleos89 committed:
Fix disabled effects signage condition

aleos89 committed:
Implement new Prontera zone for renewal
* The zone is the same as the normal Village but with included disablement of the following skills: ALL_ASSISTANT_BUYING, ALL_ASSISTANT_VENDING, ALL_BUYINGSTORE, and MC_VENDING.

aleos89 committed:
Use inheritance for Siege, WoE:TE and Velum items

aleos89 committed:
Block ALL_BUYING_STORE on Prontera

aleos89 committed:
Add some renewal maps to zones

Atemo committed:
Merge branch 'master' into feature/map_zones

Atemo committed:
Fixed conflicts

Atemo committed:
Fixed git merge hallucinations
